dev/com.ibm.ws.jca_fat/test-applications/fvtweb/src/web/JCAFVTServlet.java

```diff
@@ -732,6 +745,92 @@ public void testMBeanPurgeTwoResourceRef(HttpServletRequest request, HttpServlet
         }
     }
 
+    /**
+     * Test race condition between an invalid state change and pool purge setting state to stale
+     */
+    public void testMBeanPurgeRaceCondition(HttpServletRequest request, HttpServletResponse response) throws Throwable {
+        ConnectionFactory cf1 = (ConnectionFactory) new InitialContext().lookup("java:comp/env/jms/cf1-unshareable");
+        Connection con1 = null;
+        UserTransaction tran = (UserTransaction) new InitialContext().lookup("java:comp/UserTransaction");
+        try {
+            tran.begin();//Starting this removes the LTC
+            con1 = cf1.createConnection();
+
+            MBeanServer mbs = ManagementFactory.getPlatformMBeanServer();
+            ObjectInstance bean = getMBeanObjectInstance("jms/cf1");
+
+            con1.close();
+            tran.commit();
+
+            //Attempt to mark connections stale right before they are purged
+            ExecutorService exec = Executors.newFixedThreadPool(2);
+            List<Callable<Boolean>> callables = Arrays.asList(new MarkStale(mbs, bean), new PurgeTask(mbs, bean));
+            List<Future<Boolean>> futures = exec.invokeAll(callables);
+            awaitTerminationAfterShutdown(exec);
+
+            System.out.println("**  Pool contents after purge:\n" + getPoolContents(bean));
+            if (getPoolSize(bean) != 0)
+                throw new Exception("Not all connections were purged from the pool!");
+
+            assertTrue(futures.get(0).get().booleanValue());
+            assertTrue(futures.get(1).get().booleanValue());
+        } finally {
+            if (con1 != null)
+                con1.close();
+        }
+    }
+
+    @SuppressWarnings("restriction")
+    private class MarkStale implements Callable<Boolean> {
+        MCWrapper[] wrappers;
+
+        MarkStale(MBeanServer mbs, ObjectInstance bean) {
+            //Get at MCWrappers from poolmanager]
+            try {
+                // TODO figure out a way to get at the pool manager so that we can call:
+//                wrappers = pm.getUnSharedPoolConnections(); //NOTE this will get a lock
+            } catch (Exception e) {
+                throw new RuntimeException("Failed to get wrappers", e);
+            }
+        }
```

Comment:
Need to figure out a way to get at the PoolManager so that I can get at the connections in the unshared pool.

Or need an alternative way to change the state of the connections while purging.